File: examples/demo2.m
if ~libisloaded('mgadgetron')
    loadlibrary('mgadgetron')
end
if ~libisloaded('mutilities')
    loadlibrary('mutilities')
end

try
    input_data = gadgetron.MR_Acquisitions('testdata.h5');
    
    gadget1 = gadgetron.Gadget('RemoveROOversamplingGadget');
	gadget2 = gadgetron.Gadget('AcquisitionAccumulateTriggerGadget');
	gadget3 = gadgetron.Gadget('BucketToBufferGadget');
	gadget4 = gadgetron.Gadget('SimpleReconGadget');
	gadget5 = gadgetron.Gadget('ImageArraySplitGadget');
	gadget6 = gadgetron.Gadget('ExtractGadget');
	
    recon = gadgetron.ImagesReconstructor();

    recon.add_gadget('g1', gadget1);
	recon.add_gadget('g2', gadget2);
	recon.add_gadget('g3', gadget3);
	recon.add_gadget('g4', gadget4);
	recon.add_gadget('g5', gadget5);
    
    recon.set_input(input_data)
    recon.process()
    complex_images = recon.get_output();
    
    proc = gadgetron.ImagesProcessor();
    proc.add_gadget('g6', gadget6);
    images = proc.process(complex_images);

    for i = 1 : images.number()
        data = images.image_as_array(i);
        figure(1000000 + i)
        data = data/max(max(max(data)));
        imshow(data(:,:,1));
    end

    images.write('output2.h5', datestr(datetime))

catch err
    % display error information
    fprintf('%s\n', err.message)
    fprintf('error id is %s\n', err.identifier)
end

File: examples/demo2.py
'''
Lower-level interface demo that illustrates creating and running a chain
of gadgets and gadget sets.
'''

import argparse
import os
import sys
import time

BUILD_PATH = os.environ.get('BUILD_PATH') + '/xGadgetron'
SRC_PATH = os.environ.get('SRC_PATH') + '/xGadgetron/pGadgetron'

sys.path.append(BUILD_PATH)
sys.path.append(SRC_PATH)

from pGadgetron import *

parser = argparse.ArgumentParser(description = \
'''
Lower-level interface demo that illustrates creating and running a chain
of gadgets and gadget sets.
''')
parser.add_argument\
('filename', nargs='?', default = 'testdata.h5', \
 help = 'raw data file name (default: testdata.h5)')
args = parser.parse_args()                                 

def main():

    # acquisitions will be read from an HDF file args.filename
    input_data = MR_Acquisitions(args.filename)

    # define gadgets
    gadget1 = Gadget('RemoveROOversamplingGadget')
    gadget2 = Gadget('SimpleReconGadgetSet') # set of 4 gadgets - cf. demo1.py
    gadget3 = Gadget('ExtractGadget')

    # create reconstruction object
    recon = ImagesReconstructor()

    # build gadgets chain
    recon.add_gadget('g1', gadget1)
    recon.add_gadget('g2', gadget2)
    recon.add_gadget('g3', gadget3)

    # connect to input data
    recon.set_input(input_data)
    # perform reconstruction
    recon.process()
    
    # get reconstructed images
    images = recon.get_output()

    # show reconstructed images
    images.show()

    # write images to a new group in 'output2.h5'
    # named after the current date and time
    print('appending output2.h5...')
    time_str = time.asctime()
    images.write('output2.h5', time_str)

try:
    main()
    print('done')

except error as err:
    # display error information
    print ('Gadgetron exception occured:\n', err.value)

